galaxy/api/download/views.py

```diff
@@ -0,0 +1,80 @@
+from pathlib import Path, PurePath
+
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured, SuspiciousFileOperation
+from django.core.files.storage import FileSystemStorage
+from django.http import FileResponse
+from django.shortcuts import redirect, get_object_or_404
+from storages.backends import s3boto3
+from rest_framework.exceptions import NotFound
+from rest_framework.permissions import AllowAny
+
+from galaxy.api.base import APIView
+from galaxy.common.schema import CollectionFilename
+from galaxy.main import models
+
+
+__all__ = (
+    'ArtifactDownloadView',
+)
+
+
+class ArtifactDownloadView(APIView):
+    permission_classes = (AllowAny,)
+
+    def get(self, request, filename):
+        try:
+            filename = CollectionFilename.parse(filename)
+        except ValueError:
+            raise NotFound
+
+        version = get_object_or_404(
+            models.CollectionVersion,
+            collection__namespace__name__iexact=filename.namespace,
+            collection__name__iexact=filename.name,
+            version__exact=filename.version,
+        )
+
+        user_agent = request.META.get('HTTP_USER_AGENT', '')
+        if user_agent.startswith('Mazer/'):
+            version.collection.inc_download_count()
+
+        ca = version.get_content_artifact()
+        filename = PurePath(ca.relative_path).name
+
+        # If artifact url is a remote URL (e.g. S3 URL) return redirect to
+        # # this URL.
+        artifact_file = ca.artifact.file
+        storage = artifact_file.storage
+        if isinstance(storage, s3boto3.S3Boto3Storage):
+            return self._s3_redirect(artifact_file, filename)
+
+        # If running in debug mode serve file directly by Django.
+        if isinstance(storage, FileSystemStorage) and settings.DEBUG:
+            return self._serve_file(artifact_file.url, filename)
+
+        raise ImproperlyConfigured(
+            'Artifacts should not be served by django application'
+            ' in production mode.')
```

Comment:
Is an exception that only gets raised on production a good idea?

Another failure mode could be to serve up the file so things actually work, but log an error to alert us to the problem to correct it.

Reply:
This exception makes sense to me.

Reply:
Sounds reasonable to me. Will fix.

Reply:
Done.